_posts/2023-04-19-making-verification-compelling-visual-verification-feedback-for-dafny.markdown

```diff
@@ -0,0 +1,253 @@
+
+* **A circle with a check mark for partially proved assertions.** [Figure 1](#fig_static_icons_dark_light), line 11 and line 18.
+    While designing the icons, we got access to not only errors and declarations, <img src="/blog/assets/images/verification-compelling/example1.png" alt="Dafny gutter icons on dark theme" width="45%" style="float:left;width:25px;height:35px;object-fit:none;object-position:-7px -588px;margin:3px"/>  but also to all the individual assertions. When Dafny finds an assertion it cannot prove, it will report it, assume it, and relaunch the verification process, up to five times per declaration. Therefore if, before the fifth time, it was able to prove that a declaration is correct, it means that every assertion not tagged with an error is partially proved.
+    This is especially useful in the context of verification debugging, when one tries to copy and rewrite assertions towards the top of the declaration, manually applying the [weakest precondition calculus](https://dafny.org/dafny/DafnyRef/DafnyRef#sec-verification-debugging).
+    Not surprisingly, it solved the issue of users wondering if Dafny stopped at the first error or was able to prove the remaining. I also decided to not create an equivalent of this icon in a verified context, to keep the fully verified feedback as simple as possible.
+
+#### Dynamic icons
+
+Having icons for describing the verification status statically is great, but I found a clear benefit of using dynamic icons when displaying feedback. Typically, after looking at the verification diagnostics, users write more code or specifications, and then wait for verification to happen again. Verification can sometimes take dozens of seconds for a single file. so waiting for the entire verification of a file before updating icons would definitely not provide a good user experience...
+
+I could just have removed verification icons while Dafny is verifying declarations. You can thank me, I did not chose this easy path, because I think that this would cause not only a disturbing blinking effect when verification is fast, but also make the user not benefit from immediate comparison with the previous status. In the [weakest precondition calculus](https://dafny.org/dafny/DafnyRef/DafnyRef#sec-verification-debugging), I found it is helpful to see the red rectangle to "move up", as it often means that someone is bringing an error closer to the hypotheses. Erasing the intermediate feedback would put more strain on vision processing because users would have to visually re-align the icons and the (usually indented) code.
+
+<div id="fig_dynamic_gutter_icons"></div>
+[<img src="/blog/assets/images/verification-compelling/example-dynamic.png" alt="Dafny dynamic gutter icons" style="display:block;margin-left:auto;margin-right:auto;width:500px;max-width:95%;"/>](/blog/assets/images/verification-compelling/example-dynamic.png)
+<div style="text-align:center;font-style:italic">
```

Comment:
To aid the textual description of the animated zig-zag lines, can we make this image animated too?